mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/MongoDBDialect.java

```diff
@@ -1378,15 +1373,64 @@ private static void flushInserts(Map<DBCollection, BatchInsertionTask> inserts)
 	}
 
 	private static WriteConcern getWriteConcern(TupleContext tupleContext) {
-		return tupleContext.getOptionsContext().getUnique( WriteConcernOption.class );
+		return tupleContext.getTupleTypeContext().getOptionsContext().getUnique( WriteConcernOption.class );
 	}
 
 	private static WriteConcern getWriteConcern(AssociationContext associationContext) {
 		return associationContext.getAssociationTypeContext().getOptionsContext().getUnique( WriteConcernOption.class );
 	}
 
-	private static ReadPreference getReadPreference(TupleContext tupleContext) {
-		return tupleContext.getOptionsContext().getUnique( ReadPreferenceOption.class );
+	@SuppressWarnings("deprecation")
+	private static WriteConcern mergeWriteConcern(WriteConcern original, WriteConcern writeConcern) {
```

Comment:
I added a comment to explain the rationale behind this: the idea is that each separate operation might have write concern requirements and we need to build the stricter write concern for the aggregated operation.